File: api/app/schemas/shop.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.household import StockItem, StockItemDetail, TeaRef


class ShopRef(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    slug: str
    name: str


class ShopSummary(ShopRef):
    city: str | None
    country: str | None
    website: str | None
    image_url: str | None
    is_approved: bool
    listing_count: int


class ShopDetail(ShopSummary):
    description: str | None
    address: str | None
    created_at: datetime


class ShopCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    website: str | None = Field(default=None, max_length=500)
    address: str | None = Field(default=None, max_length=300)
    city: str | None = Field(default=None, max_length=120)
    country: str | None = Field(default=None, max_length=60)
    description: str | None = None
    image_url: str | None = Field(default=None, max_length=500)

    @model_validator(mode="after")
    def reachable_somehow(self) -> "ShopCreate":
        # Mirrors ck_shop_reachable_somehow. A shop with no website and no location is
        # not findable, and the caller deserves a 422 naming the problem rather than a
        # 500 from an IntegrityError.
        if not (self.website or self.address or self.city):
            raise ValueError("a shop needs a website, an address, or at least a city")
        return self


class ShopUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=160)
    website: str | None = Field(default=None, max_length=500)
    address: str | None = Field(default=None, max_length=300)
    city: str | None = Field(default=None, max_length=120)
    country: str | None = Field(default=None, max_length=60)
    description: str | None = None
    image_url: str | None = Field(default=None, max_length=500)


class Listing(BaseModel):
    id: uuid.UUID
    tea: TeaRef
    pack_grams: float | None
    # Integer minor units plus a currency code. A float price drifts, and a price with
    # no currency is a number, not a price.
    price_minor: int | None
    currency: str | None
    product_url: str | None
    is_available: bool


class ListingWithShop(Listing):
    shop: ShopSummary


class ListingCreate(BaseModel):
    tea_id: uuid.UUID
    pack_grams: float | None = Field(default=None, gt=0, le=999999)
    price_minor: int | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    product_url: str | None = Field(default=None, max_length=500)
    is_available: bool = True

    @model_validator(mode="after")
    def price_needs_a_currency(self) -> "ListingCreate":
        if self.price_minor is not None and self.currency is None:
            raise ValueError("currency is required when price_minor is given")
        return self


class ListingUpdate(BaseModel):
    pack_grams: float | None = Field(default=None, gt=0, le=999999)
    price_minor: int | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    product_url: str | None = Field(default=None, max_length=500)
    is_available: bool | None = None


class BuyRequest(BaseModel):
    """Recording a purchase, not taking one. No money moves through this app."""

    household_id: uuid.UUID
    grams: float = Field(gt=0, le=999999)
    price_paid_minor: int | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    purchased_at: date | None = None

    @model_validator(mode="after")
    def price_needs_a_currency(self) -> "BuyRequest":
        if self.price_paid_minor is not None and self.currency is None:
            raise ValueError("currency is required when price_paid_minor is given")
        return self


class UploadedImage(BaseModel):
    url: str


def shop_summary(shop: object, listing_count: int) -> ShopSummary:
    return ShopSummary(
        id=shop.id,  # type: ignore[attr-defined]
        slug=shop.slug,  # type: ignore[attr-defined]
        name=shop.name,  # type: ignore[attr-defined]
        city=shop.city,  # type: ignore[attr-defined]
        country=shop.country,  # type: ignore[attr-defined]
        website=shop.website,  # type: ignore[attr-defined]
        image_url=shop.image_url,  # type: ignore[attr-defined]
        is_approved=shop.is_approved,  # type: ignore[attr-defined]
        listing_count=listing_count,
    )


def shop_detail(shop: object, listing_count: int) -> ShopDetail:
    return ShopDetail(
        **shop_summary(shop, listing_count).model_dump(),
        description=shop.description,  # type: ignore[attr-defined]
        address=shop.address,  # type: ignore[attr-defined]
        created_at=shop.created_at,  # type: ignore[attr-defined]
    )


def listing_out(row: object) -> Listing:
    return Listing(
        id=row.id,  # type: ignore[attr-defined]
        tea=TeaRef.model_validate(row.tea),  # type: ignore[attr-defined]
        pack_grams=float(row.pack_grams) if row.pack_grams is not None else None,  # type: ignore[attr-defined]
        price_minor=row.price_minor,  # type: ignore[attr-defined]
        currency=row.currency,  # type: ignore[attr-defined]
        product_url=row.product_url,  # type: ignore[attr-defined]
        is_available=row.is_available,  # type: ignore[attr-defined]
    )


def listing_with_shop(row: object, listing_count: int) -> ListingWithShop:
    return ListingWithShop(
        **listing_out(row).model_dump(),
        shop=shop_summary(row.shop, listing_count),  # type: ignore[attr-defined]
    )


# Resolves StockItem.shop, a forward reference for the same reason TeaDetail.my_review is.
StockItem.model_rebuild()
StockItemDetail.model_rebuild()

File: web/src/lib/shop.ts
/**
 * The M6 shops + buying contract.
 *
 * Data shapes are aliases of the API's own OpenAPI schemas, so a renamed Pydantic field
 * breaks this build rather than a component at runtime.
 *
 * Two create bodies stay hand-written. `ListingInput.is_available` and the optional
 * fields on `ShopInput` are generated as required or as `T | null`, because
 * openapi-typescript renders a field carrying a `default` as non-optional — right for a
 * response, wrong for a request body, where the default exists so the client may omit
 * the field. `undefined` is also what `JSON.stringify` drops, which `null` is not.
 *
 * Prices are integer minor units plus a currency (3400 + "PLN" = 34.00 zł). Never render
 * `price_minor` directly; `formatPrice` exists for that.
 */
import type { components } from './generated/api'

/* --------------------------------------------------------------------------- shops */

/** The slice of a shop that hangs off something else — a tin, a listing's parent. Not
 *  `ShopSummary`: a tin does not need the shop's listing count, and embedding one would
 *  make every shelf response pay for a count nobody reads there. */
export type ShopRef = components['schemas']['ShopRef']

export type ShopSummary = components['schemas']['ShopSummary']

export type ShopDetail = components['schemas']['ShopDetail']

/* ------------------------------------------------------------------------ listings */

/**
 * One tea, in one shop, at one pack size.
 *
 * `price_minor` is an integer in the currency's minor units — 450 with `currency: 'PLN'`
 * is 4.50 zł. It is never rendered raw; `formatListingPrice` is the only thing that
 * should touch it.
 */
export type Listing = components['schemas']['Listing']

/** The same listing, seen from the tea rather than from the shop — which is why the shop
 *  travels with it. `GET /catalog/teas/{slug}/shops` is the "where to buy this" call. */
export type ListingWithShop = components['schemas']['ListingWithShop']

/* ------------------------------------------------------------------ write payloads */

/**
 * Body of both `POST /shops` (a suggestion, comes back unapproved) and
 * `POST /admin/shops` (comes back approved). Same shape, as with teas.
 *
 * The server takes `image_url` on either, but only the admin form offers a way to set
 * one — see `toShopInput`'s `withImage`. That is a product decision, not a contract
 * limit: a suggestion is six fields typed in a hurry by somebody who has not been vetted,
 * and handing every signed-in visitor a file endpoint to store bytes through is a
 * separate decision from letting them nominate a shop. The admin adds the photo when
 * they approve it.
 */
export type ShopInput = {
  name: string
  website?: string
  address?: string
  city?: string
  country?: string
  description?: string
  image_url?: string
}

/**
 * `PATCH /admin/shops/{id}`. Nullable rather than optional-only on every field that can
 * be empty, because clearing a shop's website has to be expressible as something other
 * than "leave it alone" — the same reason the generated tea and stock patches are
 * `T | null`.
 */
export type ShopPatch = components['schemas']['ShopUpdate']

export type ListingInput = {
  tea_id: string
  pack_grams?: number
  price_minor?: number
  currency?: string
  product_url?: string
  is_available?: boolean
}

export type ListingPatch = components['schemas']['ListingUpdate']

/**
 * `POST /shops/{slug}/listings/{listing_id}/buy`.
 *
 * `household_id` is on the body rather than the path because the shop, not the
 * household, is what you are standing in front of when you press the button. The server
 * answers 403 if you are not a member of it.
 *
 * `price_paid_minor` is minor units like the listing's own price: the form asks for
 * "4.50" and multiplies, because nobody types 450 for four zloty fifty.
 */
export type BuyInput = components['schemas']['BuyRequest']

/** `POST /uploads/image`, multipart, field name `file`. */
export type UploadedImage = components['schemas']['UploadedImage']

/* ------------------------------------------------------------------- query strings */

export type ShopListParams = {
  q?: string
  city?: string
  country?: string
  page?: number
  size?: number
}

export type AdminShopListParams = {
  /** A real `false` for the queue, not an omitted parameter — see the note in `toQuery`
   *  about why empty values are dropped but `false` survives. */
  approved?: boolean
  page?: number
  size?: number
}

export type ListingListParams = {
  page?: number
  size?: number
}
